ug: Add documentation for aktualizr-lite command line interface

detsch opened the pull request:
CLI documentation update for v95.

---
# PR Template and Checklist

Please complete as much as possible to speed up the reviewing process.

Readiness and adding reviewers as appropriate is required.

All PRs should be reviewed by a technical writer/documentation team and a peer.
If effecting customers—which is a majority of content changes—a member of Customer Success must also review.

## Readiness

* [ ] Merge (pending reviews)
* [ ] Merge after _date or event_
* [ ] Draft

## Overview

_Why merge this PR? What does it solve?_

## Checklist

* [x] Run spelling and grammar check, preferably with linter.
* [x] Avoid changing any header associated with a link/reference.
* [ ] Step through instructions (or ask someone to do so).
* [ ] Review for [wordiness](https://languagetool.org/insights/post/wordiness/)
* [ ] Match tone and style of page/section.
* [ ] Run `make linkcheck`.
* [ ] View HTML in a browser to check rendering.
* [ ] Use [semantic newlines](https://bobheadxi.dev/semantic-line-breaks/).
* [ ] follow best practices for commits.
  * [ ] Descriptive title written in the imperative.
  * [ ] Include brief overview of QA steps taken.
  * [ ] Mention any related issues numbers.
  * [ ] End message with sign off/DCO line (`-s, --signoff`).
  * [ ] Sign commit with your gpg key (`-S, --gpg-sign`).
  * [ ] Squash commits if needed.
* [ ] Request PR review by a technical writer and at least one peer.

## Comments

_Any thing else that a maintainer/reviewer should know._
_This could include potential issues, rational for approach, etc._

mike-sul commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,198 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+.. highlight:: bash
+   :linenothreshold: 1
+
+Sample bash script illustrating usage of CLI operations and return codes handling::
```
I would move this script to the `sotactl` repo and reference it from this doc.

detsch replied:
I was thinking about having it on a repo an also keep it in the docs, I believe it is simply enough to contribute to the docs context.

mike-sul commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,198 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+    - Unable to fetch TUF metadata
+- *14*: Failure
+    - TUF metadata not found in the provided path (offline mode only)
+- *15*: Failure
```
I think we need to ask someone from CS team to check if the CLI commands and return codes as explained here works for them.

vanmaegima commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,198 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+* ``run`` - finalizes the installation of the Target; confirms an update after reboot on a new rootfs version and/or starts the updated apps. A ``finalize`` alias is provided for this command for backwards compatibility.
+* ``update`` - performs a complete update process, including the ``check``, ``pull`` and ``install`` operations. The ``run`` operation still needs to be manually invoked after the reboot.
+
+The commands set exit codes (``echo $?``) that can be used by the caller to act accordingly.
```
This can be a new heading, similar to what we have with offline updates.

detsch replied:
Done.

vanmaegima commented:
I'm missing a note about stopping aklite daemon before running CLI.
We could cover:
- manual updates (stop aklite and stop fioconfig)
- skip daemon start from lmp-device-register
- skip daemon from the build (we have an example for custom-sota, what would be the alternative for CLI?)

vanmaegima commented:
I haven't reviewed return codes and sample script yet.

detsch commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,198 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+The `aktualizr-lite` daemon executable can be invoked to perform individual operations allowing more control over the update flow.
+This interface is subject to change over the next releases.
+
+.. prompt:: bash
```
Changed.

detsch commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,198 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+Command Line Interface (experimental)
+-------------------------------------
+The `aktualizr-lite` daemon executable can be invoked to perform individual operations allowing more control over the update flow.
+This interface is subject to change over the next releases.
```
Updated format to bold

doanac commented:
Docs for 4564d81bbd36101ef653fd0fe339153bff190d83 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2717/docs/artifacts/html/index.html

doanac commented:
Docs for 5e568db8bb424cebb1d1ce6084a4bf23e83025d1 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2742/docs/artifacts/html/index.html

vanmaegima commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,255 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+
+.. warning:: The Command Line Interface is on beta stage, and is subject to change over the next releases.
+
+.. note:: In order to use the run individual `aktualizr-lite` commands, the `aktualizr-lite` service needs to be stopped with ``sudo systemctl stop aktualizr-lite`` and/or disabled with ``sudo systemctl disable aktualizr-lite``
```
I wonder if we need to also stop fioconfig in this case as iiuc it also starts aklite daemon

detsch replied:
Yes, but perhaps we should adjust this behavior, otherwise it is a pain to use fioconfig with any custom OTA solution. I'm checkin what can we do about it, probably adjusting the `fioconfig` toml updates handler.

vanmaegima replied:
awesome, thanks!

detsch replied:
It is now merged, disabling aklite service is now enough. Fioconfig won't (re)start it anymore, even after changes in device config, such as apps list.

doanac commented:
Docs for 15489180d213f5cd85017a9b9fc156605f26699a are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2743/docs/artifacts/html/index.html

mike-sul commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,255 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+Apply Update
+^^^^^^^^^^^^
+
+The ``update`` command applies the latest available update to the device, after updating the TUF metadata and pulling the required data.
```
This `pulling the required data` is a bit confusing. IMO, it can be removed.

detsch replied:
Removed.

mike-sul commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,255 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+Apply Update
+^^^^^^^^^^^^
+
+The ``update`` command applies the latest available update to the device, after updating the TUF metadata and pulling the required data.
```
Maybe "command pulls and installs" instead of "applies" which doesn't not necessary implies downloading.

detsch replied:
Updated.

mike-sul commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,255 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+Fetch TUF Metadata and List Updates
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``check`` command will refresh the Targets metadata from the OTA server, and present you with a list of available Targets::
```
@kprosise Do you think this "refresh the Targets metadata from the OTA server" means what the command actually does?

mike-sul commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,255 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+
+   sudo aktualizr-lite update --update-name <build_number_or_name>
+
+A reboot command will be required after applying an update, followed by the execution on the  ``run`` command to finalize the update process::
```
I think the term "apply" is not used consistently across this doc. In one case it means an overall update process, in another like this it means installation.

detsch replied:
Updated.

mike-sul commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,255 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+
+.. note::
+   An update can only be performed when the original and update Targets are under the same tag.
+   In case the update is tagged differently, it is required to switch tags before running this command.
```
`the update` --> `the Target`?

detsch replied:
Updated.

doanac commented:
Docs for 9eebdc64277c4bc58fce8eac4700081894bfc394 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2751/docs/artifacts/html/index.html

detsch commented:
> I'm missing a note about stopping aklite daemon before running CLI. We could cover:
> 
> * manual updates (stop aklite and stop fioconfig)
Added note.

> * skip daemon start from lmp-device-register
Added note.

> * skip daemon from the build (we have an example for custom-sota, what would be the alternative for CLI?)
We do need the daemon binary, so the build itself is not changed someone is going to use CLI. Perhaps adding something about not enabling the aklite service by default in the build? Not sure hot that can be done, will have to check it.

doanac commented:
Docs for 968cbeef7cbe4976706db1a0f6869a173a489a44 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2752/docs/artifacts/html/index.html

mike-sul commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,217 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+.. note::
+   An update can only be performed when the original and update Targets are under the same tag.
+   In case the Target is tagged differently,
+   it is required to switch tags before running this command.
```
I would rather recommend to add the device's tag to the Target, switching tags on a device is rather the last resort.

Maybe we should remove this note at all since this is a general rule, not specific to the CLI, is applicable to any types of update.

detsch replied:
I agree. Having it here only adds uneccessary complexity to the documentation. The target would not be listed in the `list` and `check` commands anyway. Removing the note.

mike-sul reviewed (approved):
LGTM, just minor comments.

doanac commented:
Docs for f227e917db8d0b406d3e6c2e657a04a4501d9788 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2753/docs/artifacts/html/index.html

doanac commented:
Docs for c9b613078f1d8efd6375abf6db3e1dab203f25de are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2758/docs/artifacts/html/index.html

detsch commented:
@kprosise can you take a look? Feel free to do text adjustments and merge if no changes are required from our side. Thanks!

kprosise reviewed (commented):
Made some minor suggestions, but otherwise it LGTM

kprosise commented:
@detsch, if you could add the suggestions, if they seem appropriate, then squash the commits, I will then go ahead and merge. Thank you!

detsch commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The individual command line interface operations,
+specially ``check``, ``pull``, ``install`` and ``run``,
```
Applied.

detsch commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+Exit Codes
+^^^^^^^^^^
+The commands set exit codes (``echo $?``) that can be used by the caller to act accordingly.
+The possible return codes for the CLI commands are listed bellow:
```
Applied.

detsch commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+as detailed in the next section.
+
+Exit Codes
+^^^^^^^^^^
+The commands set exit codes (``echo $?``) that can be used by the caller to act accordingly.
```
Applied.

detsch commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+The ``install``  command installs the Target, which should have been previously pulled.
+It yields an error if the specified Target has not been pulled before, and also supports the ``--update-name`` option.
+
+It is necessary to verify the return codes for each command in order to guarantee the correct update process flow,
```
Applied.

detsch commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+   Only choose to do so mindfully.
+   For any update, always test before rolling out to production devices.
+
+Aktualizr-lite command line interface also allows the update steps to be performed individually,
```
Applied.

detsch commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+
+   sudo aktualizr-lite update
+
+To update to a specific build number or target name,
```
Applied.

detsch commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+
+    sudo aktualizr-lite status
+
+Fetch TUF Metadata and List Updates
```
Applied.

detsch commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+
+.. note:: If lmp-device-register is used,
+    the `--start-daemon 0` is recommended
+    in order to avoid starting aktualizr-lite daemon automatically.
```
Applied.

detsch commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+    and/or disabled with ``sudo systemctl disable aktualizr-lite``.
+
+.. note:: If lmp-device-register is used,
+    the `--start-daemon 0` is recommended
```
Applied.

detsch commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+    and is subject to change over the next releases.
+
+.. note:: In order to use the run individual `aktualizr-lite` commands,
+    the `aktualizr-lite` service needs to be stopped with ``sudo systemctl stop aktualizr-lite``
```
Applied.

detsch commented on source/user-guide/custom-sota-client.rst:
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
+- *130*: Failure
+    - Installation failed and rollback operation was not successful
+
+Automating the use of CLI operations
```
Applied.

detsch commented:
Thanks @kprosise! Applied all your suggestions, and rebased to the latest `main` . Feel free to merge when appropriate.

doanac commented:
Docs for 9d9037ddd92af8a25048bfab8a839080a23ad29d are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2764/docs/artifacts/html/index.html
